File: src/workstream_manager.py
"""
Workstream Manager - Configuration-based team grouping system
Implements Power BI-like SWITCH logic for team member categorization
"""

import json
import logging
from pathlib import Path
from typing import Dict, List, Optional, Set

logger = logging.getLogger(__name__)


class WorkstreamManager:
    """Manages workstream configuration and team member grouping"""

    def __init__(self, config_path: Optional[str] = None):
        """Initialize with workstream configuration"""
        if config_path is None:
            config_path = (
                Path(__file__).parent.parent / "config" / "workstream_config.json"
            )

        self.config_path = Path(config_path)
        self.config = self._load_config()
        self.workstreams = self.config.get("workstreams", {})
        self.default_workstream = self.config.get("default_workstream", "Others")
        self.matching_options = self.config.get("matching_options", {})

        logger.info(
            f"Loaded {len(self.workstreams)} workstreams from {self.config_path}"
        )
        logger.debug(f"Workstreams: {list(self.workstreams.keys())}")

    def _load_config(self) -> Dict:
        """Load workstream configuration from JSON file"""
        try:
            with open(self.config_path, "r") as f:
                config = json.load(f)
            logger.info(
                f"Successfully loaded workstream config from {self.config_path}"
            )
            return config
        except FileNotFoundError:
            logger.warning(
                f"Workstream config not found at {self.config_path}, using empty config"
            )
            return {"workstreams": {}, "default_workstream": "Others"}
        except json.JSONDecodeError as e:
            logger.error(f"Invalid JSON in workstream config: {e}")
            return {"workstreams": {}, "default_workstream": "Others"}

    def get_workstream_for_member(self, assigned_to: str) -> str:
        """
        Get workstream for a team member using CONTAINSSTRING logic
        Implements Power BI SWITCH(TRUE(), CONTAINSSTRING(...)) pattern
        """
        if not assigned_to or assigned_to.strip() == "":
            return self.default_workstream

        # Get matching options
        case_sensitive = self.matching_options.get("case_sensitive", False)
        partial_match = self.matching_options.get("partial_match", True)

        # Normalize name for comparison
        search_name = assigned_to if case_sensitive else assigned_to.lower()

        # Check each workstream (like SWITCH conditions)
        for workstream_name, workstream_config in self.workstreams.items():
            patterns = workstream_config.get("name_patterns", [])

            for pattern in patterns:
                # Normalize pattern for comparison
                search_pattern = pattern if case_sensitive else pattern.lower()

                # Implement CONTAINSSTRING logic
                if partial_match:
                    if search_pattern in search_name:
                        logger.debug(
                            f"Matched '{assigned_to}' to '{workstream_name}' via pattern '{pattern}'"
                        )
                        return workstream_name
                else:
                    if search_pattern == search_name:
                        logger.debug(
                            f"Matched '{assigned_to}' to '{workstream_name}' via exact pattern '{pattern}'"
                        )
                        return workstream_name

        # No match found - return default (like "Others" in Power BI)
        logger.debug(
            f"No workstream match for '{assigned_to}', using default: {self.default_workstream}"
        )
        return self.default_workstream

    def get_members_by_workstream(self, work_items: List[Dict]) -> Dict[str, List[str]]:
        """Group team members by workstream"""
        workstream_members = {}

        for item in work_items:
            assigned_to = item.get("assigned_to", "")
            workstream = self.get_workstream_for_member(assigned_to)

            if workstream not in workstream_members:
                workstream_members[workstream] = []

            if assigned_to not in workstream_members[workstream]:
                workstream_members[workstream].append(assigned_to)

        return workstream_members

    def get_workstream_members(self, workstream_name: str) -> List[str]:
        """Get all potential members for a workstream based on patterns"""
        if workstream_name not in self.workstreams:
            return []

        return self.workstreams[workstream_name].get("name_patterns", [])

    def get_available_workstreams(self) -> List[str]:
        """Get list of all configured workstreams"""
        return list(self.workstreams.keys()) + [self.default_workstream]

    def filter_work_items_by_workstream(
        self, work_items: List[Dict], workstream_names: List[str]
    ) -> List[Dict]:
        """Filter work items to only include specified workstreams"""
        if not workstream_names:
            return work_items

        filtered_items = []
        for item in work_items:
            assigned_to = item.get("assigned_to", "")
            item_workstream = self.get_workstream_for_member(assigned_to)

            if item_workstream in workstream_names:
                filtered_items.append(item)

        return filtered_items

    def get_workstream_summary(self, work_items: List[Dict]) -> Dict:
        """Generate summary of workstream distribution"""
        workstream_counts = {}
        total_items = len(work_items)

        for item in work_items:
            assigned_to = item.get("assigned_to", "")
            workstream = self.get_workstream_for_member(assigned_to)
            workstream_counts[workstream] = workstream_counts.get(workstream, 0) + 1

        # Calculate percentages
        summary = {}
        for workstream, count in workstream_counts.items():
            percentage = (count / total_items * 100) if total_items > 0 else 0
            summary[workstream] = {
                "count": count,
                "percentage": round(percentage, 1),
                "description": self.workstreams.get(workstream, {}).get(
                    "description", ""
                ),
            }

        return summary

    def validate_config(self) -> Dict[str, List[str]]:
        """Validate workstream configuration and return any issues"""
        issues = {"errors": [], "warnings": [], "info": []}

        if not self.workstreams:
            issues["warnings"].append("No workstreams configured")
            return issues

        # Check for duplicate patterns across workstreams
        all_patterns = {}
        for workstream_name, config in self.workstreams.items():
            patterns = config.get("name_patterns", [])

            if not patterns:
                issues["warnings"].append(
                    f"Workstream '{workstream_name}' has no name patterns"
                )
                continue

            for pattern in patterns:
                if pattern in all_patterns:
                    issues["errors"].append(
                        f"Pattern '{pattern}' is duplicated in workstreams: "
                        f"'{all_patterns[pattern]}' and '{workstream_name}'"
                    )
                else:
                    all_patterns[pattern] = workstream_name

        # Info about configuration
        issues["info"].append(f"Configured {len(self.workstreams)} workstreams")
        issues["info"].append(f"Total patterns: {len(all_patterns)}")
        issues["info"].append(f"Default workstream: {self.default_workstream}")

        return issues


def main():
    """Test workstream manager functionality"""
    manager = WorkstreamManager()

    # Test team member assignments
    test_members = [
        "Nenissa Malibago",
        "Christopher Jan Riños",
        "Glizzel Ann Artates",
        "Sharon Smith",
        "Unknown Person",
        "Apollo Bodiongan",
    ]

    print("=== Workstream Assignment Test ===")
    for member in test_members:
        workstream = manager.get_workstream_for_member(member)
        print(f"{member:25} -> {workstream}")

    print(f"\n=== Available Workstreams ===")
    for ws in manager.get_available_workstreams():
        print(f"- {ws}")

    print(f"\n=== Configuration Validation ===")
    validation = manager.validate_config()
    for level, messages in validation.items():
        if messages:
            print(f"{level.upper()}:")
            for msg in messages:
                print(f"  - {msg}")


if __name__ == "__main__":
    main()

File: js/workstream-manager.js
/**
 * JavaScript WorkstreamManager - Browser-first workstream filtering
 *
 * Implements Power BI-like CONTAINSSTRING logic for team member grouping.
 * This is a direct port of the Python WorkstreamManager class to JavaScript.
 */

class WorkstreamManager {
    constructor() {
        this.configuration = null;
        this.configLoaded = false;
        this.defaultConfig = this._getDefaultConfiguration();
    }

    /**
     * Initialize workstream manager with configuration
     */
    async init() {
        try {
            await this.loadConfiguration();
        } catch (error) {
            console.warn('Failed to load workstream configuration, using defaults:', error.message);
            this.configuration = this.defaultConfig;
            this.configLoaded = true;
        }
    }

    /**
     * Load workstream configuration from JSON file with fallback
     */
    async loadConfiguration() {
        try {
            // Try to load from config file first
            const response = await fetch('./config/workstream_config.json');
            if (response.ok) {
                this.configuration = await response.json();
                console.log('Workstream configuration loaded from JSON file');
            } else {
                throw new Error(`HTTP ${response.status}: ${response.statusText}`);
            }
        } catch (error) {
            console.warn('Failed to load workstream_config.json:', error.message);

            // Fallback to embedded configuration
            if (window.WORKSTREAM_CONFIG) {
                this.configuration = window.WORKSTREAM_CONFIG;
                console.log('Using embedded workstream configuration');
            } else {
                // Ultimate fallback to default configuration
                this.configuration = this.defaultConfig;
                console.log('Using default workstream configuration');
            }
        }

        this.configLoaded = true;
        this._validateConfiguration();
    }

    /**
     * Get workstream assignment for a team member (CONTAINSSTRING logic)
     * This implements the same logic as Power BI SWITCH(TRUE(), CONTAINSSTRING(...))
     */
    getWorkstreamForMember(memberName) {
        if (!this.configLoaded || !this.configuration) {
            console.warn('Workstream configuration not loaded, using default assignment');
            return this.defaultConfig.default_workstream;
        }

        if (!memberName || typeof memberName !== 'string') {
            return this.configuration.default_workstream;
        }

        const config = this.configuration;
        const options = config.matching_options || {};

        // Apply case sensitivity option
        const searchName = options.case_sensitive ?
            memberName : memberName.toLowerCase();

        // Iterate through workstreams (like SWITCH in Power BI)
        for (const [workstreamName, workstreamData] of Object.entries(config.workstreams)) {
            const namePatterns = workstreamData.name_patterns || [];

            // Check each pattern (like CONTAINSSTRING in Power BI)
            for (const pattern of namePatterns) {
                const searchPattern = options.case_sensitive ?
                    pattern : pattern.toLowerCase();

                // CONTAINSSTRING equivalent - check if pattern is contained in name
                if (searchName.includes(searchPattern)) {
                    return workstreamName;
                }
            }
        }

        // No match found - return default (like final "Others" in Power BI SWITCH)
        return config.default_workstream;
    }

    /**
     * Filter team metrics by selected workstreams
     */
    filterTeamMetrics(teamMetrics, selectedWorkstreams) {
        if (!teamMetrics || typeof teamMetrics !== 'object') {
            return {};
        }

        // If no workstreams selected or "All Teams" is selected, return all
        if (!selectedWorkstreams ||
            selectedWorkstreams.length === 0 ||
            selectedWorkstreams.includes('All Teams')) {
            return teamMetrics;
        }

        const filtered = {};

        for (const [memberName, metrics] of Object.entries(teamMetrics)) {
            const workstream = this.getWorkstreamForMember(memberName);

            if (selectedWorkstreams.includes(workstream)) {
                filtered[memberName] = metrics;
            }
        }

        return filtered;
    }

    /**
     * Get available workstreams from configuration
     */
    getAvailableWorkstreams() {
        if (!this.configLoaded || !this.configuration) {
            return ['Data', 'QA', 'OutSystems', 'Others'];
        }

        const workstreams = Object.keys(this.configuration.workstreams);

        // Add default workstream if not already present
        if (!workstreams.includes(this.configuration.default_workstream)) {
            workstreams.push(this.configuration.default_workstream);
        }

        return workstreams.sort();
    }

    /**
     * Get workstream summary for a dataset
     */
    getWorkstreamSummary(workItems) {
        if (!Array.isArray(workItems)) {
            return {};
        }

        const summary = {};
        const total = workItems.length;

        // Initialize summary for all workstreams
        for (const workstream of this.getAvailableWorkstreams()) {
            summary[workstream] = {
                count: 0,
                percentage: 0,
                description: this._getWorkstreamDescription(workstream)
            };
        }

        // Count assignments
        for (const item of workItems) {
            const assignedTo = item.assigned_to || item['Assigned To'] || '';
            const workstream = this.getWorkstreamForMember(assignedTo);

            if (summary[workstream]) {
                summary[workstream].count++;
            }
        }

        // Calculate percentages
        for (const workstream of Object.keys(summary)) {
            if (total > 0) {
                summary[workstream].percentage = (summary[workstream].count / total) * 100;
            }
        }

        return summary;
    }

    /**
     * Filter work items by workstreams
     */
    filterWorkItemsByWorkstream(workItems, selectedWorkstreams) {
        if (!Array.isArray(workItems)) {
            return [];
        }

        if (!selectedWorkstreams ||
            selectedWorkstreams.length === 0 ||
            selectedWorkstreams.includes('All Teams')) {
            return workItems;
        }

        return workItems.filter(item => {
            const assignedTo = item.assigned_to || item['Assigned To'] || '';
            const workstream = this.getWorkstreamForMember(assignedTo);
            return selectedWorkstreams.includes(workstream);
        });
    }

    /**
     * Get team members grouped by workstream from current dataset
     */
    getMembersByWorkstream(workItems) {
        if (!Array.isArray(workItems)) {
            return {};
        }

        const membersByWorkstream = {};

        // Initialize groups
        for (const workstream of this.getAvailableWorkstreams()) {
            membersByWorkstream[workstream] = [];
        }

        // Group members
        const seenMembers = new Set();

        for (const item of workItems) {
            const assignedTo = item.assigned_to || item['Assigned To'] || '';

            if (assignedTo && !seenMembers.has(assignedTo)) {
                const workstream = this.getWorkstreamForMember(assignedTo);
                membersByWorkstream[workstream].push(assignedTo);
                seenMembers.add(assignedTo);
            }
        }

        // Sort members within each workstream
        for (const workstream of Object.keys(membersByWorkstream)) {
            membersByWorkstream[workstream].sort();
        }

        return membersByWorkstream;
    }

    /**
     * Validate current configuration
     */
    validateConfiguration() {
        const validation = {
            valid: true,
            errors: [],
            warnings: []
        };

        if (!this.configuration) {
            validation.valid = false;
            validation.errors.push('No configuration loaded');
            return validation;
        }

        // Check required fields
        if (!this.configuration.workstreams) {
            validation.valid = false;
            validation.errors.push('Missing workstreams configuration');
        }

        if (!this.configuration.default_workstream) {
            validation.valid = false;
            validation.errors.push('Missing default_workstream configuration');
        }

        // Check workstream structure
        if (this.configuration.workstreams) {
            for (const [name, config] of Object.entries(this.configuration.workstreams)) {
                if (!config.name_patterns || !Array.isArray(config.name_patterns)) {
                    validation.valid = false;
                    validation.errors.push(`Workstream '${name}' missing or invalid name_patterns`);
                }

                if (config.name_patterns && config.name_patterns.length === 0) {
                    validation.warnings.push(`Workstream '${name}' has no name patterns defined`);
                }
            }
        }

        return validation;
    }

    /**
     * Get workstream description
     */
    _getWorkstreamDescription(workstream) {
        if (!this.configuration || !this.configuration.workstreams) {
            return '';
        }

        const workstreamConfig = this.configuration.workstreams[workstream];
        return workstreamConfig ? (workstreamConfig.description || '') : '';
    }

    /**
     * Validate configuration
     */
    _validateConfiguration() {
        const validation = this.validateConfiguration();

        if (!validation.valid) {
            console.error('Workstream configuration validation failed:', validation.errors);
        }

        if (validation.warnings.length > 0) {
            console.warn('Workstream configuration warnings:', validation.warnings);
        }

        return validation.valid;
    }

    /**
     * Get default configuration as fallback
     */
    _getDefaultConfiguration() {
        return {
            workstreams: {
                "Data": {
                    "description": "Data analytics and engineering team",
                    "name_patterns": [
                        "Nenissa", "Ariel", "Patrick Oniel", "Kennedy Oliveira",
                        "Christopher Jan", "Jegs", "Ian Belmonte"
                    ]
                },
                "QA": {
                    "description": "Quality assurance and testing team",
                    "name_patterns": ["Sharon", "Lorenz", "Arvin"]
                },
                "OutSystems": {
                    "description": "OutSystems development team",
                    "name_patterns": [
                        "Apollo", "Glizzel", "Prince", "Patrick Russel",
                        "Rio", "Nymar"
                    ]
                }
            },
            "default_workstream": "Others",
            "matching_options": {
                "case_sensitive": false,
                "partial_match": true,
                "match_full_name": false
            }
        };
    }
}

// Export for use in dashboards
window.WorkstreamManager = WorkstreamManager;
